lib/blue_bird/generator.ex

```diff
@@ -137,6 +146,31 @@ defmodule BlueBird.Generator do
     )
   end
 
+  @spec controllers([%PhxRoute{}]) :: [atom]
+  defp controllers(routes) do
+    Enum.reduce(routes, [], fn(route, list) ->
+      [Module.concat([:Elixir | Module.split(route.plug)]) | list]
+    end) |> Enum.uniq
+  end
+
+  @spec extract_groups([atom]) :: Map.t
+  defp extract_groups(list, list \\ %{})
+  defp extract_groups([], groups) do
+    groups
+  end
+  defp extract_groups([controller | list], groups) do
+    %{name: name, description: description} = apply(controller, :api_group, [])
+    extract_groups(
+      list,
+      Map.put(
+        groups, name, %Group{name: name, description: description}
```

Comment:
This will put out a map like this:

```
%{"Bobtails" => %BlueBird.Group{description: "The Bobtail Resource",
   name: "Bobtails"}}
```

So we have the name as a key and at the same time as a value in a struct. We don't gain anything by doing it this way. I suggest to simplify this to a simple map with the name as the key and the description as the value:

```
%{"Bobtails" => "The Bobtail Resource"}
```

We wouldn't even need the `%BlueBird.Group{}` struct and could simply define the group field of the `%ApiDoc{}` struct as a `map`.

Reply:
I thought about adding `:note`, `:warning` and even `:parameter` later.

Reply:
I just discussed this with @rhazdon. We would like to keep the apigroup macro small like it is. We'd rather not add more fields/parameters to it.

The notes and warnings are rendered as ordinary markdown, so you could just add them to the description. A reason to allow them to be defined outside of the description in routes would be that we could extend BlueBird so that it would be possible to define common notes and warnings in a central place and just reference them within the macro. We don't see an advantage in doing the same on group level.

As for parameters, yes, it should be possible to define common parameters centrally, but we don't think that the apigroup macro is the right place for this.

And that's why we think that the solution suggested earlier should be the way to do it. 

We're going to merge your PR with the other outstanding PR and release it later. Thanks a lot!

Reply:
Please let us know when you're finished with your PR, then we'll merge.